File: JavaScript/729-My Calendar I.js
var TreeNode = function(start, end){
    this.left = null;
    this.right = null;
    this.start = start;
    this.end = end;
}

var MyCalendar = function() {
    this.root = null;
};

/** 
 * @param {number} start 
 * @param {number} end
 * @return {boolean}
 */
MyCalendar.prototype.book = function(start, end) {
    let insert = (start, end) =>{
        let head = this.root;
        while (true){
            if (head.end <= start){
                if (!head.right){
                    head.right = new TreeNode(start, end);
                    return true;
                }
                head = head.right;
            }else if (head.start >= end){
                if (!head.left){
                    head.left = new TreeNode(start, end);
                    return true;
                }
                head = head.left;
            }else{
                return false;
            }
        }
    }
    
    if (!this.root){
        this.root = new TreeNode(start, end);
        return true;
    }
    
    return insert(start, end);
};

/** 
 * Your MyCalendar object will be instantiated and called as such:
 * var obj = new MyCalendar()
 * var param_1 = obj.book(start,end)
 */

File: C++/729-My Calendar I.cpp
class MyCalendar {
public:
    set<pair<int, int>> s; // multiset also works here
    MyCalendar() {
        
    }
    
    bool book(int start, int end) {
        auto it = s.upper_bound({end, -1});
        if (it == s.begin()) {  // note that when a set is empty, s.begin() and s.end() are equal
            s.insert({start, end});
            return true;
        }
        advance(it, -1);
        int preStart = it->first, preEnd = it->second;
        if (start >= preEnd) {
            s.insert({start, end});
            return true;
        }
        return false;
    }
};

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar* obj = new MyCalendar();
 * bool param_1 = obj->book(start,end);
 */